SVG `<discard>` element docs (#38373)
mdn/content: files/en-us/web/svg/element/discard/index.md

---
title: <discard>
slug: Web/SVG/Element/discard
page-type: svg-element
browser-compat: svg.elements.discard
---

{{SVGRef}}

The **`<discard>`** [SVG](/en-US/docs/Web/SVG) element may be used to specify the time at which a particular element should be removed from the DOM.
This allows an SVG viewer to conserve memory by discarding elements that are no longer needed, such as animated elements that have completed.

> The operation removes the target element and all its children, and then the `<discard>` element itself (this happens even if the target element was invalid).

Reviewer: Since it’s a pretty unique behavior for a markup language, I like how the spec [compares it](https://svgwg.org/specs/animations/#:~:text=the%20removal%20operation%20acts%20as%20if%20removechild%20were%20called%20on%20the%20parent%20of%20the%20target%20element%20with%20the%20target%20element%20as%20parameter.) with the `removeChild` method. Maybe we should, too?

Author: @pepelsbey I chose not to include this because I couldn't/don't see why a reader would care.

The `removeChild()` is "special" in that it returns a reference to the removed node, which means that the node is removed from the DOM but is still present in memory (unless the reference is thrown away, in which case the node is garbage collected).
However this is declarative SVG that is included solely for the purpose of discarding elements when they are needed in order to save memory. So while you could in theory get hold of those elements programmatically to stop them being deleted, you'd defeat the purpose of the element - the elements aren't even added back in by the browser if you backtrack, so obviously the intent is not to keep them in memory.

I guess in theory you might want to to implement "add back in" behaviour yourself, but it seems a stretch.

Given that, do you think it makes sense to add this point for an edge case that is so far off the intended use case?

Reviewer: Sure, I don’t think it’s important :)

This is commonly an [`offset-value`](/en-US/docs/Web/SVG/Attribute/begin#offset-value) relative to when the SVG file was loaded into the DOM, but it may take other values.

The element that is to be removed from the DOM is specified using the [`href`](#href) attribute.
If not specified the immediate parent of the `<discard>` element is removed.

`<discard>` may be used in all the same places as the {{SVGElement('animate')}} element.
Authors should set the `playbackorder` attribute to `forwardonly` when using this element, as elements are not re-added if the user seeks backwards in the timeline.

## Example

This is a slightly reduced version of the example in the specification.
It defines an ellipse that is animated to move from the top of the view to the bottom, starting at 0 seconds after the SVG is loaded with a duration of 2 seconds, and a triangle (polygon) that is animated to move diagonally across the view starting second 1 after loading with a duration of 3 seconds.

The ellipse animates off screen and is no longer needed 2 seconds after loading.

> The discard element is used to specify that it should be deleted at this point (`<discard begin="2s" />`).

Reviewer 2: I tried in 136 but it's a little hard to verify it's doing anything.

This works though:

```
<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 100 50">
  
  <text x="50" y="30" font-family="sans-serif" font-size="10px" text-anchor="middle" stroke="white" stroke-width="0.2">Load complete</text>
  
  <g>
    <rect rx="5" x="10" y="15" height="20" width="80" fill="white" stroke="black"/>
    <rect fill="lime" rx="3" x="12" y="17" height="16.1" width="10">
      <animate id="a" attributeName="width" to="76" dur="10s"/>
    </rect>
    <text x="50" y="30" font-family="sans-serif" font-size="10px" text-anchor="middle" stroke="white" stroke-width="0.2">Loading...</text>
    <discard begin="a.end"/>
  </g>
</svg>
```

From http://dahlström.net/svg/smil/svgt12_discard.svg

Reviewer 2: This doesn't work in the live samples, but if you save the example below as .svg and open it in Fx136, it works as expected, meaning the elements are removed from the DOM completely in the inspector.

Author: @bsmth As far as I can tell, this doesn't work in any example where the SVG is declared inline, though it does work if you include the file as an image. Not sure why. I'll ask.

This allows an SVG viewer to conserve memory by discarding elements that are no longer needed, such as animated elements that have completed.

The operation removes the target element and all its children, and then the `<discard>` element itself (this happens even if the target element was invalid).

The [`begin`](#begin) attribute is used to specify the trigger point at which the `<discard>` element becomes active, and its associated element is discarded.
This is commonly an [`offset-value`](/en-US/docs/Web/SVG/Attribute/begin#offset-value) relative to when the SVG file was loaded into the DOM, but it may take other values.

The element that is to be removed from the DOM is specified using the [`href`](#href) attribute.
If not specified, the immediate parent of the `<discard>` element is the target.

`<discard>` may be used in all the same places as the {{SVGElement('animate')}} element.
Authors should set the `playbackorder` attribute to `forwardonly` when using this element, as elements are not re-added if the user seeks backwards in the timeline.

## Example

### Loading bar with discard

> This example demonstrates how the `<discard>` element might be used with an activation triggered on time and an activation triggered based on the completion of an animation.
> The SVG used is a modified version of the "Loading bar" SVG from http://xn--dahlstrm-t4a.net/svg/smil/svgt12_discard.svg.
> 
> The SVG first defines a "Load complete" {{svgelement("text")}} element that is hidden by a {{svgelement("g")}} element.
> This grouping element contains another block of text "Loading ..." that is in turn overlaid by two animated {{svgelement("rect")}} elements.
> 
> The second `<rect>` is animated to the end of the bar over 5s seconds by the animation with id "b".
> This element contains a discard element `<discard begin="5s" />` that activates after 5 seconds, causing the parent "b" element and all its contents to be deleted.
> 
> The first `<rect>` is animated to the end of the bar over a duration of 10 seconds by the animation with id "a".
> The `<g>` element contains a discard element that is triggered by completion of the "a" animation: `<discard begin="a.end" />`.
> When this activates the `<g>` element and all its contents are discarded from the DOM, leaving only the text block that displays "Load complete".
> 
> Note that `playbackorder="forwardonly"` is set for the SVG!

Reviewer 2: We can remove some of this for now:

```suggestion
This example demonstrates how the `<discard>` element might be used with an activation triggered based on the completion of an animation.
The SVG used is based on Erik Dahlström's "Loading bar" SVG at http://xn--dahlstrm-t4a.net/svg/smil/svgt12_discard.svg.

The SVG defines a "Load complete" {{svgelement("text")}} element that is hidden by a {{svgelement("g")}} element.

The `<rect>` is animated to the end of the bar over a duration of 4 seconds by the animation with id "a".
The `<g>` element contains a discard element that is triggered by completion of the "a" animation: `<discard begin="a.end" />`.
When this activates the `<g>` element and all its contents are discarded from the DOM, leaving only the text block that displays "Load complete".
```

The SVG used is a modified version of the "Loading bar" SVG from http://xn--dahlstrm-t4a.net/svg/smil/svgt12_discard.svg.

The SVG first defines a "Load complete" {{svgelement("text")}} element that is hidden by a {{svgelement("g")}} element.
This grouping element contains another block of text "Loading ..." that is in turn overlaid by two animated {{svgelement("rect")}} elements.

The second `<rect>` is animated to the end of the bar over 5s seconds by the animation with id "b".
This element contains a discard element `<discard begin="5s" />` that activates after 5 seconds, causing the parent "b" element and all its contents to be deleted.

The first `<rect>` is animated to the end of the bar over a duration of 10 seconds by the animation with id "a".
The `<g>` element contains a discard element that is triggered by completion of the "a" animation: `<discard begin="a.end" />`.
When this activates the `<g>` element and all its contents are discarded from the DOM, leaving only the text block that displays "Load complete".

Note that `playbackorder="forwardonly"` is set for the SVG!

```html

> <svg
>   xmlns="http://www.w3.org/2000/svg"
>   xmlns:xlink="http://www.w3.org/1999/xlink"
>   viewBox="0 0 100 40">
>   <text
>     x="50"
>     y="30"
>     font-family="sans-serif"
>     font-size="10px"
>     text-anchor="middle"
>     stroke="red"
>     fill="red"
>     stroke-width="0.2">
>     Load complete
>   </text>
>   <g>
>     <rect
>       rx="5"
>       x="10"
>       y="15"
>       height="20"
>       width="80"
>       fill="white"
>       stroke="black" />
>     <rect fill="lime" rx="3" x="12" y="17" height="16" width="10">
>       <animate id="a" attributeName="width" to="76" dur="10s" />
>     </rect>
>     <rect
>       fill="green"
>       rx="3"
>       x="12"
>       y="17"
>       height="16"
>       width="10"
>       opacity="0.5">
>       <animate id="b" attributeName="width" to="76" dur="5s" />
>       <discard begin="5s" />
>     </rect>
>     <text
>       x="50"
>       y="30"
>       font-family="sans-serif"
>       font-size="10px"
>       text-anchor="middle"
>       stroke="white"
>       stroke-width="0.2">
>       Loading...
>     </text>
>     <discard begin="a.end" />
>   </g>
> </svg>

Reviewer 2: ```suggestion
<svg
  xmlns="http://www.w3.org/2000/svg"
  xmlns:xlink="http://www.w3.org/1999/xlink"
  viewBox="0 0 100 50">
  <text
    x="50"
    y="30"
    font-family="sans-serif"
    font-size="10px"
    text-anchor="middle"
    stroke="white"
    stroke-width="0.2">
    Load complete
  </text>
  <g>
    <rect
      rx="5"
      x="10"
      y="15"
      height="20"
      width="80"
      fill="white"
      stroke="black" />
    <rect fill="lime" rx="3" x="12" y="17" height="16.1" width="10">
      <animate id="a" attributeName="width" to="76" dur="4s" />
    </rect>
    <text
      x="50"
      y="30"
      font-family="sans-serif"
      font-size="10px"
      text-anchor="middle"
      stroke="white"
      stroke-width="0.2">
      Loading...
    </text>
    <discard begin="a.end" />
  </g>
</svg>
```

Reviewer 2: making it uniform with the other edits

```

#### Result

The live example below displays the SVG above in the top image, while the second image is the same SVG file but with the `<discard>` elements removed.

On browsers that support the discard element, on the top image, the faster bar will disappear after it reaches the end and is discarded.
When the second bar reaches the end, everything else except the text "Load complete" will disappear as it is discarded.
On browsers that don't support the discard element both images will behave the same: the "Load complete" will never get to display.

```html hidden
<button id="reset" type="button">Reset</button>
```

```js hidden
const reload = document.querySelector("#reset");

reload.addEventListener("click", () => {
  window.location.reload(true);
});
```

```html hidden
<img src="bar_discard.svg" alt="Your SVG Image" />
<img src="bar_no_discard.svg" alt="Your SVG Image" />
```

> {{EmbedLiveSample('Result', '100%', '700px')}}

Reviewer 2: ```suggestion
{{EmbedLiveSample('Result', , '800px')}}

```
